File: packages/core/src/observability/approval-artifact.ts
/**
 * Protected identity and validation for deterministic approval artifacts.
 *
 * @internal
 * @module
 */

import { z } from "zod";

import { sha256Hex } from "../content/sha256";
import { canonicalEvidenceJson } from "../evidence/canonical-json";
import type {
  CruxArtifactId,
  CruxArtifactKind,
  CruxAttributes,
  CruxRunId,
} from "./contract";

const APPROVAL_PREFIX = "approval_";
const MAX_APPROVAL_SCALARS = 512;
const MAX_APPROVAL_BYTES = 2_048;
const UTF8_ENCODER = new TextEncoder();

/** Stable occurrence identity shared by request and decision production. */
export interface ApprovalArtifactOccurrence {
  readonly domain: "crux.tool.approval";
  readonly identityEpoch: 1;
  readonly namespace: {
    readonly operationId: CruxRunId;
    readonly runId: CruxRunId;
  };
  readonly approvalId: string;
  readonly slot: "request" | "decision";
}

/** Exact protected attributes carried by one deterministic approval artifact. */
export type ApprovalArtifactAttributes = CruxAttributes & {
  readonly approvalOccurrence: ApprovalArtifactOccurrence;
};

export const ApprovalArtifactAttributesSchema = z
  .object({
    approvalOccurrence: z
      .object({
        domain: z.literal("crux.tool.approval"),
        identityEpoch: z.literal(1),
        namespace: z
          .object({
            operationId: z.string().min(1),
            runId: z.string().min(1),
          })
          .strict(),
        approvalId: z.string().refine(isApprovalId, {
          message: "Invalid tool approval identity",
        }),
        slot: z.enum(["request", "decision"]),
      })
      .strict(),
  })
  .strict();

/** Construct a detached protected occurrence marker. */
export function approvalArtifactAttributes(
  occurrence: ApprovalArtifactOccurrence,
): ApprovalArtifactAttributes {
  const parsed = ApprovalArtifactAttributesSchema.parse({
    approvalOccurrence: occurrence,
  });
  return Object.freeze({
    approvalOccurrence: Object.freeze({
      ...parsed.approvalOccurrence,
      namespace: Object.freeze({
        operationId:
          parsed.approvalOccurrence.namespace.operationId as CruxRunId,
        runId: parsed.approvalOccurrence.namespace.runId as CruxRunId,
      }),
    }),
  });
}

/** Derive the opaque full-SHA-256 artifact identity from the exact marker. */
export function approvalArtifactId(
  attributes: ApprovalArtifactAttributes,
): CruxArtifactId {
  const canonical = canonicalEvidenceJson(attributes);
  return `artifact_${sha256Hex(UTF8_ENCODER.encode(canonical))}` as CruxArtifactId;
}

/** Validate the public approval protocol ID without Unicode normalization. */
export function isApprovalId(value: string): boolean {
  if (
    !value.startsWith(APPROVAL_PREFIX) ||
    hasUnpairedSurrogate(value) ||
    [...value].length > MAX_APPROVAL_SCALARS ||
    UTF8_ENCODER.encode(value).byteLength > MAX_APPROVAL_BYTES
  ) {
    return false;
  }
  const suffix = value.slice(APPROVAL_PREFIX.length);
  const scalars = [...suffix];
  return (
    scalars.length > 0 &&
    !scalars.some(isControlCharacter) &&
    !isApprovalBoundaryWhitespace(scalars[0]!) &&
    !isApprovalBoundaryWhitespace(scalars.at(-1)!)
  );
}

interface ApprovalArtifactRecordShape {
  readonly operationId: string;
  readonly runId: string;
  readonly artifactId: string;
  readonly kind: CruxArtifactKind;
  readonly attributes?: Readonly<Record<string, unknown>>;
}

/** Enforce marker, kind/slot, namespace, and derived artifact identity. */
export function validateApprovalArtifact(
  artifact: ApprovalArtifactRecordShape,
  context: z.RefinementCtx,
): void {
  const isApprovalKind =
    artifact.kind === "approval.request" ||
    artifact.kind === "approval.decision";
  const hasMarker =
    artifact.attributes !== undefined &&
    Object.hasOwn(artifact.attributes, "approvalOccurrence");
  if (!isApprovalKind && !hasMarker) return;

  const parsed = ApprovalArtifactAttributesSchema.safeParse(
    artifact.attributes,
  );
  if (!parsed.success) {
    for (const issue of parsed.error.issues) {
      context.addIssue({
        code: "custom",
        message: issue.message,
        path: ["attributes", ...issue.path],
      });
    }
    return;
  }
  const occurrence = parsed.data.approvalOccurrence;
  const expectedKind =
    occurrence.slot === "request"
      ? "approval.request"
      : "approval.decision";
  if (artifact.kind !== expectedKind) addIssue(context, "kind");
  if (
    occurrence.slot === "request" &&
    (occurrence.namespace.operationId !== artifact.operationId ||
      occurrence.namespace.runId !== artifact.runId)
  ) {
    addIssue(context, "attributes");
  }
  if (
    artifact.artifactId !==
    approvalArtifactId(parsed.data as ApprovalArtifactAttributes)
  ) {
    addIssue(context, "artifactId");
  }
}

function hasUnpairedSurrogate(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      if (index + 1 >= value.length) return true;
      const next = value.charCodeAt(index + 1);
      if (next < 0xdc00 || next > 0xdfff) return true;
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return true;
    }
  }
  return false;
}

function isControlCharacter(value: string): boolean {
  const code = value.codePointAt(0)!;
  return (code >= 0 && code <= 0x1f) || (code >= 0x7f && code <= 0x9f);
}

function isApprovalBoundaryWhitespace(value: string): boolean {
  const code = value.codePointAt(0)!;
  return (
    (code >= 0x09 && code <= 0x0d) ||
    code === 0x20 ||
    code === 0x85 ||
    code === 0xa0 ||
    code === 0x1680 ||
    (code >= 0x2000 && code <= 0x200a) ||
    (code >= 0x2028 && code <= 0x2029) ||
    code === 0x202f ||
    code === 0x205f ||
    code === 0x3000 ||
    code === 0xfeff
  );
}

function addIssue(context: z.RefinementCtx, field: string): void {
  context.addIssue({
    code: "custom",
    message: "Invalid protected tool approval artifact",
    path: [field],
  });
}

File: packages/local/internal/observability/approval_artifact.go
package observability

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"strings"
	"unicode/utf8"
)

const (
	approvalIDPrefix     = "approval_"
	maxApprovalIDScalars = 512
	maxApprovalIDBytes   = 2_048
)

type approvalArtifactAttributes struct {
	ApprovalOccurrence approvalArtifactOccurrence `json:"approvalOccurrence"`
}

type approvalArtifactOccurrence struct {
	Domain        string                    `json:"domain"`
	IdentityEpoch int                       `json:"identityEpoch"`
	Namespace     approvalArtifactNamespace `json:"namespace"`
	ApprovalID    string                    `json:"approvalId"`
	Slot          string                    `json:"slot"`
}

type approvalArtifactNamespace struct {
	OperationID string `json:"operationId"`
	RunID       string `json:"runId"`
}

// validateApprovalArtifact recognizes and validates the exact protected
// occurrence marker used by deterministic tool-approval artifacts.
func validateApprovalArtifact(artifact ArtifactRecord) (bool, error) {
	isApprovalKind := artifact.Kind == "approval.request" ||
		artifact.Kind == "approval.decision"
	hasMarker, err := hasApprovalOccurrenceMarker(artifact.Attributes)
	if err != nil {
		return isApprovalKind, err
	}
	if !isApprovalKind && !hasMarker {
		return false, nil
	}
	if !hasMarker {
		return true, fmt.Errorf("approval artifact marker is required")
	}
	attributes, err := decodeApprovalArtifactAttributes(artifact.Attributes)
	if err != nil {
		return true, err
	}
	occurrence := attributes.ApprovalOccurrence
	if occurrence.Domain != "crux.tool.approval" ||
		occurrence.IdentityEpoch != 1 ||
		!validApprovalID(occurrence.ApprovalID) {
		return true, fmt.Errorf("approval artifact occurrence is invalid")
	}
	expectedKind := "approval.request"
	if occurrence.Slot == "decision" {
		expectedKind = "approval.decision"
	} else if occurrence.Slot != "request" {
		return true, fmt.Errorf("approval artifact slot is invalid")
	}
	if artifact.Kind != expectedKind {
		return true, fmt.Errorf("approval artifact kind and slot do not match")
	}
	if occurrence.Namespace.OperationID == "" ||
		occurrence.Namespace.RunID == "" {
		return true, fmt.Errorf("approval artifact namespace is invalid")
	}
	if occurrence.Slot == "request" &&
		(occurrence.Namespace.OperationID != artifact.OperationID ||
			occurrence.Namespace.RunID != artifact.RunID) {
		return true, fmt.Errorf("approval request namespace does not match its envelope")
	}
	expectedID, err := approvalArtifactID(attributes)
	if err != nil {
		return true, err
	}
	if artifact.ArtifactID != expectedID {
		return true, fmt.Errorf("approval artifact id is invalid")
	}
	return true, nil
}

func decodeApprovalArtifactAttributes(
	raw json.RawMessage,
) (approvalArtifactAttributes, error) {
	if !utf8.Valid(raw) {
		return approvalArtifactAttributes{}, fmt.Errorf(
			"approval artifact marker is not valid UTF-8",
		)
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.DisallowUnknownFields()
	var attributes approvalArtifactAttributes
	if err := decoder.Decode(&attributes); err != nil {
		return approvalArtifactAttributes{}, fmt.Errorf(
			"approval artifact marker is invalid: %w",
			err,
		)
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF {
		return approvalArtifactAttributes{}, fmt.Errorf(
			"approval artifact marker contains trailing JSON",
		)
	}
	return attributes, nil
}

func hasApprovalOccurrenceMarker(raw json.RawMessage) (bool, error) {
	if len(raw) == 0 {
		return false, nil
	}
	if !utf8.Valid(raw) {
		return false, fmt.Errorf("artifact attributes are not valid UTF-8")
	}
	var probe map[string]json.RawMessage
	if err := json.Unmarshal(raw, &probe); err != nil {
		return false, fmt.Errorf("artifact attributes are invalid: %w", err)
	}
	_, marked := probe["approvalOccurrence"]
	return marked, nil
}

func approvalArtifactID(attributes approvalArtifactAttributes) (string, error) {
	canonical, err := canonicalEvidenceJSON(attributes)
	if err != nil {
		return "", fmt.Errorf("canonical approval artifact identity: %w", err)
	}
	sum := sha256.Sum256(canonical)
	return "artifact_" + hex.EncodeToString(sum[:]), nil
}

func validApprovalID(value string) bool {
	if !utf8.ValidString(value) ||
		!strings.HasPrefix(value, approvalIDPrefix) ||
		len(value) > maxApprovalIDBytes ||
		utf8.RuneCountInString(value) > maxApprovalIDScalars {
		return false
	}
	suffix := strings.TrimPrefix(value, approvalIDPrefix)
	if suffix == "" {
		return false
	}
	first, _ := utf8.DecodeRuneInString(suffix)
	last, _ := utf8.DecodeLastRuneInString(suffix)
	if approvalBoundaryWhitespace(first) || approvalBoundaryWhitespace(last) {
		return false
	}
	for _, scalar := range suffix {
		if scalar <= 0x1f || (scalar >= 0x7f && scalar <= 0x9f) {
			return false
		}
	}
	return true
}

func approvalBoundaryWhitespace(value rune) bool {
	return (value >= 0x09 && value <= 0x0d) ||
		value == 0x20 ||
		value == 0x85 ||
		value == 0xa0 ||
		value == 0x1680 ||
		(value >= 0x2000 && value <= 0x200a) ||
		(value >= 0x2028 && value <= 0x2029) ||
		value == 0x202f ||
		value == 0x205f ||
		value == 0x3000 ||
		value == 0xfeff
}
